src/metabase/models/card.clj

```diff
@@ -306,27 +341,58 @@
                            :query-database        query-db-id
                            :field-filter-database field-db-id})))))))
 
-(defn- assert-valid-model
+(defn- ensure-type-and-dataset-are-consistent
+  "We're in the process of migrating from using `report_card.dataset` to `report_card.type`.
+  In the future we'll drop `dataset` and only use `type`. But for now we need to make sure that both keys are aligned
+  when dealing with cards.
+  - If both keys are present, we make sure `dataset` is true if `type` is `model` else false.
+    This will make a different when we have `metric` type since a boolean can't represent tri-state
+  - If only one key is present, we'll assoc the correct value for the other key."
+  [{:keys [type dataset] :as card}]
+  (cond
+   ;; if none of the 2 keys is present, do nothing
+   (and (nil? type) (nil? dataset))
+   card
+
+   ;; if both type and dataset is present, we prioritize type
+   (and (some? type) (some? dataset))
+   (assoc card :dataset (= type "model"))
+
+   ;; if only type is present, make sure dataset follows
+   (some? type)
+   (assoc card :dataset (= type "model"))
+
+   ;; if only dataset is present, make sure type follows
+   (some? dataset)
+   (let [inferred-type (if dataset
+                         "model"
+                         "question")]
+     (log/warnf "Card type not found, defaulting to '%s'" inferred-type)
+     (assoc card :type inferred-type))))
+
+(defn- assert-valid-type
   "Check that the card is a valid model if being saved as one. Throw an exception if not."
-  [{:keys [dataset dataset_query]}]
-  (when dataset
-    (let [template-tag-types (->> (vals (get-in dataset_query [:native :template-tags]))
-                                  (map (comp keyword :type)))]
-      (when (some (complement #{:card :snippet}) template-tag-types)
-        (throw (ex-info (tru "A model made from a native SQL question cannot have a variable or field filter.")
-                        {:status-code 400}))))))
+  [{:keys [type dataset_query]}]
+  (case type
+    "model" (let [template-tag-types (->> (get-in dataset_query [:native :template-tags])
+                                          vals
+                                          (map (comp keyword :type)))]
+              (when (some (complement #{:card :snippet}) template-tag-types)
+                (throw (ex-info (tru "A model made from a native SQL question cannot have a variable or field filter.")
+                                {:status-code 400}))))
+    nil))
 
 ;; TODO -- consider whether we should validate the Card query when you save/update it??
 (defn- pre-insert [card]
   (let [defaults {:parameters         []
                   :parameter_mappings []}
-        card     (merge defaults card)]
+        card     (ensure-type-and-dataset-are-consistent (merge defaults card))]
```

Comment:
This should be in the `t2/define-before-insert` instead. Otherwise there could be other functions called in `t2/define-before-insert` that depend on `:dataset` or `:type` that do not receive the valid card returned by `ensure-type-and-dataset-are-consistent`.

Also should you put this in `t2/define-before-update` too? There are at least some tests that call `t2/update!` that update `:dataset` directly using toucan .

Reply:
I'm not following.

`ensure-type-and-dataset-are-consistent` is both triggered in `define-before-insert` and `define-before-update`, it's just indirectly.

Reply:
Sorry you're right, it is in `define-before-update`. I was looking at the wrong branch in my editor.

My first point is that I'd prefer `t2/define-before-insert` to include the call to `ensure-type-and-dataset-are-consistent` directly at at the beginning of the thread macro, exactly as you have done in `t2/define-before-update`. The benefit is to future-proof against accidentally depending on `:dataset` or `:type` being valid in the `define-before-insert` body before `ensure-type-and-dataset-are-consistent` has been called. For example, if `populate-result-metadata` was changed to use on `:dataset` for whatever reason, it could introduce a bug where `:dataset` is out-of-sync with `:type`.

Reply:
makes sense, moved this call to `before-insert`